lib/ecraft/extensions/mixins/hashable.rb

```diff
@@ -0,0 +1,110 @@
+#
+# Utility mixin for converting arbitrary objects to Hashes.
+#
+# Similar to https://github.com/mustafaturan/hashable, but different in implementation.
+#
+
+require 'active_support'
+require 'active_support/core_ext/hash/keys'
+
+module Ecraft
+  module Extensions
+    module Mixins
+      module Hashable
+        def to_h(*targets)
+          target = get_target(targets)
+
+          if target.is_a?(Hash)
+            strip_string_recurse_enumerable_values(target)
+          elsif target.is_a?(Array)
+            strip_array(target)
+          elsif target.is_a?(OpenStruct)
+            target.marshal_dump
+          elsif value_type?(target)
+            target
+          # We need to check to ensure we don't call ourselves recursively.
+          elsif !target.class.included_modules.include?(Hashable) && target.respond_to?(:to_h)
+            target.to_h
+          else
+            turn_object_into_hash(target)
+          end
+        end
+
+        def to_hash(target = nil)
+          to_h(target)
+        end
+
+        private
+
+        def get_target(targets)
+          if targets.empty?
+            self
+          else
+            targets.first
+          end
+        end
+
+        def value_type?(v)
+          v.is_a?(BigDecimal) ||
+            v.is_a?(Integer) ||
+            v.is_a?(Float) ||
+            v.is_a?(String) ||
+            v.is_a?(TrueClass) ||
+            v.is_a?(FalseClass)
+        end
+
+        def strip_string_recurse_enumerable_values(hash)
+          hash.each_value do |v|
+            v.rstrip! if v.respond_to?(:rstrip!) && !v.frozen?
+            to_h(v) if v.is_a?(Enumerable)
+          end
+          hash.symbolize_keys
+        end
+
+        def strip_array(array)
+          array.map do |v|
+            if v.nil?
+              nil
+            elsif v.is_a?(String)
+              v.rstrip
+            elsif v.respond_to?(:to_h) && !v.is_a?(Hash)
+              v.to_h
+            else
+              to_h(v)
+            end
+          end
+        end
+
+        # Returns a Symbol-keyed Hash based on target's instance variables.
+        #
+        # @param target [Object]
+        #
+        # @return [Hash] with Symbol keys
+        def turn_object_into_hash(target)
+          target.instance_variables.each_with_object({}) do |var, hash|
+            original = target.instance_variable_get(var)
+            value = original
+
+            if value.is_a?(Hash)
+              # Nothing to do - it's already a hash, so we presume we don't have to do do any more conversions of it.
+            elsif value.is_a?(Array) && value.first.is_a?(Hash)
+              # Nothing to do - an Array of Hashes already have a sane enough format.
+            else
+              map_enumerable(original) if original.respond_to?(:each)
+              value = original.to_h if original.respond_to?(:to_h)
+              value.rstrip! if value.respond_to?(:rstrip!)
```

Comment:
(I understand this is a literal code move, but I could not resist reading the text.)

The defensive "!value.frozen?" detail is not here. Is that intentional?

Reply:
Makes sense, we should add that. Thanks for suggesting!